### PR TITLE
feat(payment_methods_list): add is_tax_connector_enabled boolean value in payment_methods_list call response

### PR DESCRIPTION
**Pull request by ImSagnik007**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Added `skip_external_tax_calculation` boolean in payments create request that indicates whether tax to be calculated for this intent or not.
Added `is_tax_calculation_enabled` boolean value in `payment_methods_list` call response that checks both this `skip_external_tax_calculation` and `is_tax_connector_enabled` in business profile.

Payments create request:
`curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_OcQoFePRNVjYvsA9YGWKITq1BEHtInBKEtW9IRbKyBbOaBjwc6WYGwuxnyBVuhZh' \
--data-raw '{
    "amount": 1800,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "customer_id": "cus_NwicjmCJbgafjhjfhHr6",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "three_ds",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "9123456789",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "return_url": "https://duck.com",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "payment_method": "pay_later",
    "payment_method_type": "klarna",
    "payment_experience": "invoke_sdk_client",
    "payment_method_data": {
        "pay_later": {
            "klarna_sdk": {
                "token": "65bfb654-ef2c-6b39-942e-ffdc9a105d3c"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "city": "Groningen",
            "state": "Groningen",
            "zip": "94105",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "+4901761428434",
            "country_code": "+1"
        }
    },
    "shipping": {
        "address": {
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "city": "Groningen",
            "state": "Groningen",
            "zip": "94105",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "446681800",
            "country_code": "+41"
        }
    },
    "metadata": {},
    "order_details": [
        {
            "product_name": "Apple iphone 15",
            "quantity": 1,
            "amount": 180000,
            "account_name": "transaction_processing"
        }
    ],
      "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "128.0.0.1"
    },
    "skip_external_tax_calculation": true
}'`

PML call response:
`{
    "redirect_url": "https://google.com/success",
    "currency": "USD",
    "payment_methods": [
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "payment_method_type": "klarna",
                    "payment_experience": [
                        {
                            "payment_experience_type": "redirect_to_url",
                            "eligible_connectors": [
                                "klarna"
                            ]
                        },
                        {
                            "payment_experience_type": "invoke_sdk_client",
                            "eligible_connectors": [
                                "klarna"
                            ]
                        }
                    ],
                    "card_networks": null,
                    "bank_names": null,
                    "bank_debits": null,
                    "bank_transfers": null,
                    "required_fields": {
                        "billing.address.country": {
                            "required_field": "payment_method_data.billing.address.country",
                            "display_name": "country",
                            "field_type": {
                                "user_address_country": {
                                    "options": [
                                        "AU",
                                        "AT",
                                        "BE",
                                        "CA",
                                        "CZ",
                                        "DK",
                                        "FI",
                                        "FR",
                                        "DE",
                                        "GR",
                                        "IE",
                                        "IT",
                                        "NL",
                                        "NZ",
                                        "NO",
                                        "PL",
                                        "PT",
                                        "ES",
                                        "SE",
                                        "CH",
                                        "GB",
                                        "US"
                                    ]
                                }
                            },
                            "value": "US"
                        }
                    },
                    "surcharge_details": null,
                    "pm_auth_connector": null
                }
            ]
        }
    ],
    "mandate_payment": null,
    "merchant_name": "NewAge Retailer",
    "show_surcharge_breakup_screen": false,
    "payment_type": "normal",
    "request_external_three_ds_authentication": false,
    "collect_shipping_details_from_wallets": false,
    "collect_billing_details_from_wallets": false,
    "is_tax_calculation_enabled": false
}`

We are preventing to update the value of `skip_external_tax_calculation` during payment update

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by Aprabhat19 — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 1]
```diff
@@ -3547,6 +3551,13 @@ pub async fn list_payment_methods(
             business_profile.collect_billing_details_from_wallet_connector
         }));
 
+    let is_tax_connector_enabled = business_profile
+        .as_ref()
+        .map(|business_profile| business_profile.is_tax_connector_enabled)
+        .unwrap_or(false);
+
+    // let skip_external_tax_calculation = payment_intent.as_ref().and_then(|payment_intent|payment_intent.skip_external_tax_calculation);
```
[the remark this answers is not in the document]
Comment (ImSagnik007) [reply crates/router/src/core/payment_methods/cards.rs]: done

**Review by ShankarSinghC — COMMENTED** [review]
Add a PR description explaining the use case of adding this filed in the payment method list.

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -522,6 +522,8 @@ pub struct PaymentsRequest {
         example = "Custom_Order_id_123"
     )]
     pub merchant_order_reference_id: Option<String>,
+
+    pub skip_external_tax_calculation: Option<bool>,
```
Comment (Aprabhat19) [inline crates/api_models/src/payments.rs:526]: Add  appropriate  documentation above this field to be displayed in Api Reference docs

File: crates/api_models/src/payment_methods.rs [hunk 3]
```diff
@@ -1482,6 +1482,8 @@ pub struct PaymentMethodListResponse {
 
     /// flag that indicates whether to collect billing details from wallets or from the customer
     pub collect_billing_details_from_wallets: Option<bool>,
+
+    pub is_tax_calculation_enabled: bool,
```
[the remark this answers is not in the document]
Comment (ImSagnik007) [reply crates/api_models/src/payment_methods.rs:1488]: done

File: crates/api_models/src/payments.rs (hunk 2 above)
Comment (ImSagnik007) [reply crates/api_models/src/payments.rs:527]: done

**Review by ShankarSinghC — DISMISSED** [review]

**Review by Aprabhat19 — APPROVED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 4]
```diff
@@ -3590,6 +3599,7 @@ pub async fn list_payment_methods(
             request_external_three_ds_authentication,
             collect_shipping_details_from_wallets,
             collect_billing_details_from_wallets,
+            is_tax_calculation_enabled: is_tax_connector_enabled && skip_external_tax_calculation,
```
[the remark this answers is not in the document]
Comment (SamraatBansal) [reply crates/router/src/core/payment_methods/cards.rs:3602]: If you want to keep this put 

```
let skip_external_tax_calculation = payment_intent
        .as_ref()
        .and_then(|intent| intent.skip_external_tax_calculation)
        .unwrap_or(true);
```

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ImSagnik007) [reply crates/router/src/core/payment_methods/cards.rs]: Done

File: crates/router/src/core/payments/operations/payment_update.rs [hunk 5]
```diff
@@ -758,6 +758,9 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
             .payment_intent
             .merchant_order_reference_id
             .clone();
+        let skip_external_tax_calculation =
+            payment_data.payment_intent.skip_external_tax_calculation;
+
         payment_data.payment_intent = state
```
[the remark this answers is not in the document]
Comment (ImSagnik007) [reply crates/router/src/core/payments/operations/payment_update.rs]: Done

File: crates/router/src/core/payment_methods/cards.rs [hunk 6]
```diff
@@ -3506,6 +3506,10 @@ pub async fn list_payment_methods(
         });
     }
     let currency = payment_intent.as_ref().and_then(|pi| pi.currency);
+    let skip_external_tax_calculation = payment_intent
+        .as_ref()
+        .and_then(|intent| intent.skip_external_tax_calculation)
+        .unwrap_or(false);
```
Comment (SanchithHegde) [inline crates/router/src/core/payment_methods/cards.rs:3512]:
```suggestion
        .unwrap_or(true);
```

File: crates/router/src/core/payment_methods/cards.rs [hunk 7]
```diff
@@ -3550,6 +3554,11 @@ pub async fn list_payment_methods(
             business_profile.collect_billing_details_from_wallet_connector
         }));
 
+    let is_tax_connector_enabled = business_profile
+        .as_ref()
+        .map(|business_profile| business_profile.is_tax_connector_enabled)
+        .unwrap_or(false);
```
Comment (SanchithHegde) [inline crates/router/src/core/payment_methods/cards.rs:3560]: Can we have a method on the business profile domain model that considers both the `is_tax_connector_enabled` field and the `tax_connector_id` field in the business profile, that returns a boolean indicating whether tax connector is enabled or not?

Comment (ImSagnik007) [reply crates/router/src/core/payment_methods/cards.rs]: @SanchithHegde added validation for tax_connector_id

File: crates/router/src/core/payment_methods/cards.rs [hunk 8]
```diff
@@ -3649,6 +3653,14 @@ pub async fn list_payment_methods(
             business_profile.collect_billing_details_from_wallet_connector
         }));
 
+    let is_tax_connector_enabled = business_profile.as_ref().map_or(false, |business_profile| {
+        let is_tax_connector_enabled = business_profile.is_tax_connector_enabled;
+        match &business_profile.tax_connector_id {
+            Some(id) if !id.is_empty() => is_tax_connector_enabled,
+            _ => false,
+        }
+    });
```
Comment (SanchithHegde) [inline crates/router/src/core/payment_methods/cards.rs:3662]: Can we add this as a method on the business profile domain model instead?

Comment (ImSagnik007) [reply crates/router/src/core/payment_methods/cards.rs]: Done

**Review by SanchithHegde — DISMISSED** [review]

**Review by SamraatBansal — DISMISSED** [review]

**Review by ShankarSinghC — APPROVED** [review]

**Review by apoorvdixit88 — DISMISSED** [review]
Dashboard specific changes look fine

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 9]
```diff
@@ -659,6 +659,14 @@ impl From<BusinessProfileSetter> for BusinessProfile {
...
 
 impl BusinessProfile {
+    pub fn get_is_tax_connector_enabled(&self) -> bool {
+        let is_tax_connector_enabled = self.is_tax_connector_enabled;
+        match &self.tax_connector_id {
+            Some(_id) => is_tax_connector_enabled,
+            _ => false,
+        }
+    }
+
```
Comment (SamraatBansal) [inline crates/hyperswitch_domain_models/src/business_profile.rs:669]: cc: @Narayanbhat166 @SanchithHegde

**Review by SamraatBansal — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 10]
```diff
@@ -659,6 +659,14 @@ impl From<BusinessProfileSetter> for BusinessProfile {
 }
 
 impl BusinessProfile {
+    pub fn get_is_tax_connector_enabled(&self) -> bool {
+        let is_tax_connector_enabled = self.is_tax_connector_enabled;
+        match &self.tax_connector_id {
+            Some(_id) => is_tax_connector_enabled,
+            _ => false,
+        }
+    }
```
Comment (racnan) [inline crates/hyperswitch_domain_models/src/business_profile.rs:668]: [is_some_and](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_and) can probably help simplify this.

**Review by racnan — APPROVED** [review]
nitpick

File: crates/hyperswitch_domain_models/src/business_profile.rs (hunk 10 above)
Comment (ImSagnik007) [reply crates/hyperswitch_domain_models/src/business_profile.rs:668]: @racnan this function will be refactored later, while introducing tax_connector_id validation

**Pull request merged**